### DISCUSSION TITLE
Product Gallery - Full View Mode

### DISCUSSION THREAD

**Discussion by gigitux** [woocommerce/woocommerce#44952] [WooCommerce, Blocks, and Site Editing]
Currently, we are actively developing a new Product Gallery block #9891. This new block is designed to be highly customizable, empowering merchants to adjust the placement of any inner block effortlessly through the Site Editor:

https://github.com/woocommerce/woocommerce-blocks/assets/4463174/fbda3ca9-9669-4b51-a063-3924385efeb0

One of the exciting features we are working on is the Full View Mode, which enhances the user experience in the following way:

![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/af9eefcc-9bef-4e2a-93c3-b514b727664e)

This immersive mode becomes visible when a user visits a Single Product Page and clicks on the primary image within the Product Gallery.

To ensure a consistent and flexible user experience, we aim to extend the same level of customization capabilities to the Full View Mode as well. In the future, it may be necessary to enable merchants to customize the "full view mode" in a specific manner for individual products.

Given the current feature set of WP/Gutenberg, we have identified two possible solutions to achieve this:

- template parts.
- template.

### Template Parts:

> A Template part is a block type used for structuring your site. It is a container block used to display other blocks, for example, in a site header or site footer. The template part block is available in the block inserter in the template editing mode and the Site Editor.

By [this definition](https://developer.wordpress.org/themes/block-themes/templates-and-template-parts/#block-c5fa39a2-a27d-4bd2-98d0-dc6249a0801a:~:text=A%20Template%20part%20is%20a%20block%20type%20used%20for%20structuring%20your%20site.%20It%20is%20a%20container%20block%20used%20to%20display%20other%20blocks%2C%20for%20example%2C%20in%20a%20site%20header%20or%20site%20footer.%20The%20template%20part%20block%20is%20available%20in%20the%20block%20inserter%20in%20the%20template%20editing%20mode%20and%20the%20Site%20Editor.), adopting a template part could be a good approach because the Product Gallery in Full View Mode is just a modal container that wraps a "Product Gallery Block". The issues are that:
- the template part will be available in the block inserter.
- it is impossible to create a specific template part for a specific product.

### Template:

> A theme can include different types of templates: Site templates like home, archive, and 404, which are part of the [template hierarchy](https://developer.wordpress.org/themes/basics/template-hierarchy/) and can be edited in the Site Editor. And page templates that are assigned to a post or page and edited in the template editor.

By [this definition](https://developer.wordpress.org/themes/block-themes/templates-and-template-parts/#block-798d87ea-3c2f-4065-a24e-215b644acc95:~:text=A%20theme%20can%20include%20different%20types%20of%20templates%3A%20Site%20templates%20like%20home%2C%20archive%2C%20and%20404%2C%20which%20are%20part%20of%20the%20template%20hierarchy%20and%20can%20be%20edited%20in%20the%20Site%20Editor.%20And%20page%20templates%20that%20are%20assigned%20to%20a%20post%20or%20page%20and%20edited%20in%20the%20template%20editor.), the template should be used for customize pages, but the Product Gallery - Full View Mode isn't a page. At the same time:
- The template will not be visible in the block inserter.
- It is possible to create a specific template for a specific product.

I crafted a PR to illustrate the user experience using the template approach. This is the current UX:

https://github.com/woocommerce/woocommerce-blocks/assets/4463174/36e9f286-4ae6-4449-b1a8-f7188cd198c6

What do you think? Should we use the template or the template part (and maybe contribute upstream to fix the current constraints of the template parts), or should we think of another solution that doesn't involve the templates?

**Comment by passatgt** [comment]
Tell me if i'm wrong, but couldn't you just show the modal in the product page editor? When you click on the main image, the modal is visible and you can customize it in place just like the other blocks. For example if you have a navigation, you can click on a menu and open the submenu and you can interact with it to change fonts, colors etc... This is kinda the same, just not a submenu, but a modal window instead. Once you close it you can continue editing the rest of the site. This way you don't need either a template or a template part i believe.

**Comment by shaunandrews** [comment]
One thing to consider here is that the full-screen gallery can be invoked from Product Gallery blocks across many templates. For example, the Single Product page and Shop page could both show one or more Product Gallery blocks. All of these blocks should trigger the same full-screen gallery template.

**Reply by passatgt** [reply]
In that case, yes, it should be a separate template, however i don't think this is needs to be displayed anywhere else. In the loops it should be a quick view with add to cart i believe, not just an image gallery.

**Comment by Marc-pi** [comment]
seems to be a Magnific Popup style or Modal ?
could be an option in the Single Product Template, associated to the Image Gallery block, no ?

**Comment by nerrad** [comment]
I appreciate the thought going into considering how this is implemented. A few initial thoughts:

- Since you mentioned this immersive mode is attached primarily to usage within the context of a single product template, I would expect the editing of this particular view should be something triggered from the context of the single product template.
- The ability to create a specific template for this view per product is superfluous to me and I think would increase friction for users. All product-specific functionality should be enabled via a specific product-type template. So whatever solution is derived is something that ideally would be embedded with the template, so when product-specific single-product templates are created, this view would also be specific to that product.
- Within WP, templates historically have a semantic connection to the layout of what is fully rendered for a given route and fit into a hierarchy. Having this full view implemented as a template breaks that pattern.
- While semantically this seems to fit better as a template part, and template parts _can have their blocks detached_, the advantages of template parts still don't have a straightforward way to implement as an editing tool for the Full View Mode while keeping open the option of specific implementations of this attached to a product specific single product template in the future.

Given the above, it identifies that templates and template parts in their current iterations are not satisfactory containers for this mode:

- The primary behaviour is something that feels _block-specific_.
- The problem remains surfacing a focused editing view for the "Full View" content. 

Short term, it does seem like using a modal for editing this content would be a good initial approach as it would not lock us into a poorer pattern.

Long term, this feels like something that would be useful to expose as another editing view type that block authors (or even WP core) could implement. Perhaps transitioning in and out of the same focused view that is used for editing patterns and template parts for blocks that have additional "modes" to edit?

**Reply by shaunandrews** [reply]
> Since you mentioned this immersive mode is attached primarily to usage within the context of a single product template, I would expect the editing of this particular view should be something triggered from the context of the single product template.

I mention above that this gallery could be triggered from any template that makes use of a product gallery; It seems like a mistake to limit it only to the Single Product template.

**Comment by nerrad** [comment]
I have another idea:

- A template part still seems to be the best semantic representation of this while also providing a focused view for editing.
- We'd need a way to transition to the template part from the product gallery block instance (and back).
- In the future, we could enable the ability to _switch out_ what template part is attached to a given product gallery instance. The chooser would utilize template part category to restrict which template parts are available to select from.

The site editor already has a number of the elements needed for helping this work and we could implement additional affordances at the block level to aid in discovery.

@shaunandrews what do you think of this? It'd keep things attached to the block instance while still allowing for synced behaviour controlled via template parts.

**Reply by nerrad** [reply]
There's already a built-in mechanism for users to create new types of template parts:
<img width="691" alt="CleanShot 2023-09-06 at 08 06 27@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/1429108/64781983-1d65-4f7e-aa5c-c4d40ef0f1d8">

**Reply by shaunandrews** [reply]
A Template Part works for me. I also like that it would show up here:

<img width="344" alt="CleanShot 2023-09-06 at 09 21 21@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/191598/956d20c0-ff69-4672-b0b4-d5988d425635">

I wonder if it would make sense to consolidate "Store" partials, rather than have one for Mini-Cart and (now) one for Full-Screen Gallery?

--

> We'd need a way to transition to the template part from the product gallery block instance (and back).

I think we could start by adding a simple link to the setting in the Inspector:

<img width="546" alt="CleanShot 2023-09-06 at 09 22 35@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/191598/e32f41c1-2f24-4e18-80b8-8c8e466f7cb5">

For the "and back" bit, if you're using the Site Editor then (I think?) there will already be a back button. This is similar to the flow for editing a Header from within another template.

<img width="88" alt="CleanShot 2023-09-06 at 09 23 56@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/191598/da8cb2fa-70cc-40ec-b86a-8ffc98367747">

**Reply by shaunandrews** [reply]
<img width="523" alt="CleanShot 2023-09-06 at 09 26 13@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/191598/a90b41f4-988a-425d-bf26-a1068b1043b9">

We could also look to include the Full-Screen Gallery partial when listing other partials, like a Header or Footer within a template's inspector sidebar.

**Comment by nerrad** [comment]
Gutenberg is currently addressing something similar here: https://github.com/WordPress/gutenberg/issues/43852

**Comment by gigitux** [comment]
Thanks for the feedback!
Since Gutenberg is addressing a similar issue for the Navigation block (https://github.com/WordPress/gutenberg/issues/43852), we will follow the same approach! We will use a template part.

With the template part approach, we have two points to investigate:
- Transitioning from the product gallery block to the template part.
- Attaching a specific template part to a particular instance of the product gallery.

Since our current priority is not centered on providing support for specific template parts tailored to individual product gallery instances, our immediate efforts in the coming weeks will be directed toward refining the transition process from the product gallery block to the template part. 


I want to mention that for the second point, in the future we could contribute to GB: https://github.com/WordPress/gutenberg/issues/43852#issuecomment-1278669827
